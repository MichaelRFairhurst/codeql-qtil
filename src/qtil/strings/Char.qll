private import qtil.inheritance.Instance

private module Str {
  import qtil.strings.Join
}

class Char extends InfInstance<int>::Type {
  bindingset[this]
  Char() { exists(inst().toUnicode()) }

  bindingset[this]
  string toString() { result = inst().toUnicode() }

  bindingset[this]
  int codePoint() { result = inst() }

  bindingset[this]
  predicate isLowercase() { toString().isLowercase() }

  bindingset[this]
  Char toLowercase() { result = toString().toLowerCase().codePointAt(0) }

  bindingset[this]
  predicate isUppercase() { toString().isUppercase() }

  bindingset[this]
  Char toUppercase() { result = toString().toUpperCase().codePointAt(0) }

  bindingset[this]
  predicate isDigit() { toString() in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"] }

  bindingset[this]
  predicate isAlphabetic() { not toString().toLowerCase() = toString().toUpperCase() }

  bindingset[this]
  predicate isSpecial() {
    not isDigit() and
    not isAlphabetic()
  }

  bindingset[this]
  predicate isAscii() {
    codePoint() >= 0 and
    codePoint() <= 127
  }

  bindingset[s]
  predicate isStr(string s) { s.codePointAt(0) = this }

  bindingset[this, n]
  string repeat(int n) {
    n >= 0 and
    result = concat(int x | x = [1 .. n] | this.toString())
  }

  bindingset[this, s]
  string wrap(string s) { result = this.toString() + s + this.toString() }

  bindingset[this, v1, v2]
  string join(string v1, string v2) { result = Str::join(this.toString(), v1, v2) }

  bindingset[this, v1, v2, v3]
  string join(string v1, string v2, string v3) { result = Str::join(this.toString(), v1, v2, v3) }

  bindingset[this, v1, v2, v3, v4]
  string join(string v1, string v2, string v3, string v4) {
    result = Str::join(this.toString(), v1, v2, v3, v4)
  }

  bindingset[this, v1, v2, v3, v4, v5]
  string join(string v1, string v2, string v3, string v4, string v5) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5)
  }

  bindingset[this, v1, v2, v3, v4, v5, v6]
  string join(string v1, string v2, string v3, string v4, string v5, string v6) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5, v6)
  }

  bindingset[this, v1, v2, v3, v4, v5, v6, v7]
  string join(string v1, string v2, string v3, string v4, string v5, string v6, string v7) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5, v6, v7)
  }

  bindingset[this, v1, v2, v3, v4, v5, v6, v7, v8]
  string join(string v1, string v2, string v3, string v4, string v5, string v6, string v7, string v8) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5, v6, v7, v8)
  }

  bindingset[this, s, idx]
  string split(string s, int idx) {
    result = s.splitAt(this.toString(), idx)
  }

  bindingset[this, s]
  string split(string s) {
    result = s.splitAt(this.toString())
  }

  bindingset[this, s]
  int indexIn(string s) {
    result = s.indexOf(this.toString())
  }

  bindingset[this, s, startAt]
  int indexIn(string s, int idx, int startAt) {
    result = s.indexOf(this.toString(), idx, startAt)
  }
}

bindingset[s]
Char charOf(string s) { result.isStr(s) }
