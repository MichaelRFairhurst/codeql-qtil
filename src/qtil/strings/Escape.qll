/**
 * Add escape sequence (e.g., backslash) to a string before a relevant character (e.g., double
 * quote).
 *
 * For instance, `"foo%bar".escape("%", "$") = "foo$%bar"`, and
 * `"foo%bar$baz".escape("%", "$") = "foo$%bar$$baz"`.
 */
bindingset[escapee, escaper, str]
string escape(string str, string escapee, string escaper) {
  result = str.replaceAll(escaper, escaper + escaper).replaceAll(escapee, escaper + escapee)
}

/**
 * Add escape sequence (e.g., backslash) to a string before a relevant character (e.g., double
 * quote).
 *
 * For instance, `"foo%bar".escape("%") = "foo\\%bar"`, and
 * `"foo%bar\\baz".escape(",") = "foo\\%bar\\\\baz"`.
 */
bindingset[escapee, str]
string escape(string str, string escapee) { result = escape(str, escapee, "\\") }

/**
 * Wrap a string with a delimiter (e.g., double quote), adding a specifiable escape character if
 * necessary.
 *
 * For instance, `wrapEscaping("foo", "%", "$") = "%foo%"`, and
 * `wrapEscaping("foo%bar", "%", "$") = "%foo$%bar%"`.
 */
bindingset[delimiter, escape, str]
string wrapEscaping(string str, string delimiter, string escape) {
  result = delimiter + escape(str, delimiter, escape) + delimiter
}

/**
 * Wrap a string with a delimiter (e.g., double quote), adding a backslash escape character if
 * necessary.
 *
 * For instance, `wrapEscaping("foo", "%") = "%foo%"`, and
 * `wrapEscaping("foo%bar", "%") = "%foo\\%bar%"`.
 */
bindingset[delimiter, str]
string wrapEscaping(string str, string delimiter) { result = wrapEscaping(str, delimiter, "\\") }

/**
 * Wrap a string with double quotes, adding a backslash escape character if necessary.
 *
 * For instance, `doubleQuoteEscaping("foo") = "\"foo\""`, and
 * `doubleQuoteEscaping("foo\"bar") = "\"foo\\\"bar\""`.
 */
bindingset[str]
string doubleQuoteEscaping(string str) { result = wrapEscaping(str, "\"") }

/**
 * Wrap a string with single quotes, adding a backslash escape character if necessary.
 *
 * For instance, `singleQuoteEscaping("foo") = "'foo'"`, and
 * `singleQuoteEscaping("foo'bar") = "'foo\\'bar"`.
 */
bindingset[str]
string singleQuoteEscaping(string str) { result = wrapEscaping(str, "'") }

/**
 * Add escapes to a string literal to make it safe for use in a Java-style PRCE regex, as a direct
 * character by character match.
 *
 * For instance, `regexEscape("foo") = "foo"`, and
 * `regexEscape("^foo(bar)baz[qux]$") = "\\^foo\\(bar\\)baz\\[qux\\]\\$"`.
 */
bindingset[str]
string escapeRegex(string str) {
  result = str.regexpReplaceAll("[{}()\\[\\].+*?^$\\\\|]", "\\\\$0")
}

/**
 * Unescape a string by removing the specified escape character (e.g., backslash).
 *
 * For instance, `unescape("foo%bar", "%") = "foobar"`, and
 * `unescape("foo%bar%%baz", "%") = "foobar%baz"`.
 */
bindingset[escaper, str]
string unescape(string str, string escaper) {
  // Replace "${escaper}${char}" with "${char}". We must escape the escape character
  // to avoid it being interpreted as a regex special character.
  result = str.regexpReplaceAll(escapeRegex(escaper) + "(.)", "$1")
}

/**
 * Unescape a string by removing the backslash escape character.
 *
 * For instance, `unescape("foo\\%bar") = "foo%bar"`, and
 * `unescape("foo\\%bar\\\\baz") = "foo%bar\\baz"`.
 */
bindingset[str]
string unescape(string str) { result = unescape(str, "\\") }

/**
 * Get the contents of a wrapped string value, which may contain escape sequences using the
 * specified, escape sequence, removing the wrapping characters and one level of escapes.
 *
 * For instance, `unwrap("%foo%", "%", "$") = "foo"`, and
 * `unwrap("%foo$%bar$$baz%", ) = "foo%bar$baz`.
 *
 * Does not hold for any result value in the case where the string is not wrapped with the specified
 * delimiter.
 */
bindingset[str, delimiter, escape]
string unwrap(string str, string delimiter, string escape) {
  exists(string unwrapped |
    unwrapped = str.substring(delimiter.length(), str.length() - delimiter.length()) and
    unwrapped = delimiter + unwrapped + delimiter and
    result = unescape(unwrapped, escape)
  )
}

/**
 * Get the contents of a double quoted string literal, removing the double quotes and any escape
 * characters.
 *
 * For instance, `undoubleQuote("\"foo\"") = "foo"`, and
 * `undoubleQuote("\"foo\\\"bar\"") = "foo\"bar"`.
 * 
 * Does not hold for any result value in the case where the string is not double quoted.
 */
bindingset[str]
string undoubleQuote(string str) {
    result = unwrap(str, "\"", "\\")
}

/**
 * Get the contents of a single quoted string literal, removing the single quotes and any escape
 * characters.
 *
 * For instance, `unsingleQuote("'foo'") = "foo"`, and
 * `unsingleQuote("'foo\\'bar'") = "foo'bar"`.
 */
bindingset[str]
string unsingleQuote(string str) { result = unescape(str.substring(1, str.length() - 1)) }
