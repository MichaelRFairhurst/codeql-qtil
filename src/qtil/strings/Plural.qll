bindingset[str]
string plural(string str) { result = str + "s" }

bindingset[str, pcount]
string plural(string str, int pcount) { result = plural(str, plural(str), pcount) }

bindingset[single, plural, pcount]
string plural(string single, string plural, int pcount) {
  if pcount = 1 then result = single else result = plural
}
