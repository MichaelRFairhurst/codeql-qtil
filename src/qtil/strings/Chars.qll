import qtil.strings.Char

/**
 * A module that declares predicates for defining common separators to the various `ListBuilder`
 * modules defined in this file.
 */
module Chars {
  pragma[inline]
  Char comma() { result.isStr(",") }

  pragma[inline]
  Char dollar() { result.isStr("$") }

  pragma[inline]
  Char colon() { result.isStr(":") }

  pragma[inline]
  Char at() { result.isStr("@") }

  pragma[inline]
  Char hash() { result.isStr("#") }

  pragma[inline]
  Char excl() { result.isStr("!") }

  pragma[inline]
  Char caret() { result.isStr("^") }

  pragma[inline]
  Char amp() { result.isStr("&") }

  pragma[inline]
  Char pipe() { result.isStr("|") }

  pragma[inline]
  Char semicolon() { result.isStr(";") }

  pragma[inline]
  Char plus() { result.isStr("+") }

  pragma[inline]
  Char minus() { result.isStr("-") }

  pragma[inline]
  Char slash() { result.isStr("/") }

  pragma[inline]
  Char backslash() { result.isStr("\\") }

  pragma[inline]
  Char dot() { result.isStr(".") }

  pragma[inline]
  Char question() { result.isStr("?") }

  pragma[inline]
  Char percent() { result.isStr("%") }

  pragma[inline]
  Char tilde() { result.isStr("~") }

  pragma[inline]
  Char space() { result.isStr(" ") }

  pragma[inline]
  Char tab() { result.isStr("\t") }

  pragma[inline]
  Char newline() { result.isStr("\n") }

  pragma[inline]
  Char backtick() { result.isStr("`") }
}
