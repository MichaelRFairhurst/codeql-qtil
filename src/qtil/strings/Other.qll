import qtil.parameterization.SignaturePredicates
import qtil.parameterization.SignatureTypes
import qtil.strings.Char

/**
 * CodeQL 2.21.1 has a bug where `concat` with a delimiter and an order will silently drop the
 * delimiter.
 *
 * Use this module to work around that bug.
 *
 * Example:
 *
 * ```
 * string names(int x) {
 *   x = 0 and result = "foo" or
 *   x = 1 and result = "bar" or
 *   x = 2 and result = "baz"
 * }
 *
 * // BUG: selects foobarbaz
 * select concat(int x, string name | name = names(x) | name order by x, ",")
 *
 * // FIXED: selects foo,bar,baz
 * select ConcatDelimOrderFixed<Chars::comma/0, names/1>::join()
 * ```
 */
module ConcatDelimOrderFixed<
  Nullary::Ret<Char>::pred/0 separator, Unary<int>::Ret<string>::pred/1 items>
{
  /**
   * Perform the concatenation with the given separator and items/order.
   */
  string join() {
    exists(int lastIdx |
      lastIdx = max(int x | exists(items(x))) and
      result =
        concat(int x, string given, string separated |
          given = items(x) and
          if x = lastIdx then separated = given else separated = given + separator()
        |
          separated order by x
        )
    )
    or
    // Match CodeQL concat() behavior and always return an empty string if there are no items.
    not exists(items(_)) and result = ""
  }
}
