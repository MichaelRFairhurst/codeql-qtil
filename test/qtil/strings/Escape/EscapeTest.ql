import qtil.strings.Escape
import qtil.testing.Qnit

class TestEscapeNothingToDo extends Test, Case {
  override predicate run(Qnit test) {
    if escape("foo", "$", "%") = "foo"
    then test.pass("Basic escape nothing to do works")
    else test.fail("Basic escape nothing to do doesn't work")
  }
}

class TestEscapeBasic extends Test, Case {
  override predicate run(Qnit test) {
    if escape("foo$bar", "$", "%") = "foo%$bar"
    then test.pass("Basic escape works")
    else test.fail("Basic escape doesn't work")
  }
}

class TestEscapeWillEscapeTheEscaper extends Test, Case {
  override predicate run(Qnit test) {
    if escape("foo$bar$baz", "$", "%") = "foo%$bar%$baz"
    then test.pass("Escape will escape the escaper")
    else test.fail("Escape not properly escaping the escaper")
  }
}

class TestEscapeWithBackslashDefault extends Test, Case {
  override predicate run(Qnit test) {
    if escape("foo$bar", "$") = "foo\\$bar"
    then test.pass("Escape with backslash default works")
    else test.fail("Escape with backslash default doesn't work")
  }
}

class TestEscapeWithBackslashWillEscapeTheEscaper extends Test, Case {
  override predicate run(Qnit test) {
    if escape("foo$bar\\baz", "$") = "foo\\$bar\\\\baz"
    then test.pass("Escape with backslash will escape the escaper")
    else test.fail("Escape with backslash not properly escaping the escaper")
  }
}

class TestWrapEscapingNothingToDo extends Test, Case {
  override predicate run(Qnit test) {
    if wrapEscaping("foo", "$", "%") = "$foo$"
    then test.pass("Basic wrap escaping nothing to do works")
    else test.fail("Basic wrap escaping nothing to do doesn't work")
  }
}

class TestWrapEscapingBasic extends Test, Case {
  override predicate run(Qnit test) {
    if wrapEscaping("foo$bar", "$", "%") = "$foo%$bar$"
    then test.pass("Basic wrap escaping works")
    else test.fail("Basic wrap escaping doesn't work")
  }
}

class TestWrapEscapingWillEscapeTheEscaper extends Test, Case {
  override predicate run(Qnit test) {
    if wrapEscaping("foo$bar$baz", "$", "%") = "$foo%$bar%$baz$"
    then test.pass("Wrap escaping will escape the escaper")
    else test.fail("Wrap escaping not properly escaping the escaper")
  }
}

class TestWrapEscapingWithBackslashDefault extends Test, Case {
  override predicate run(Qnit test) {
    if wrapEscaping("foo$bar", "$") = "$foo\\$bar$"
    then test.pass("Wrap escaping with backslash default works")
    else test.fail("Wrap escaping with backslash default doesn't work")
  }
}

class TestDoubleQuoteEscapingNothingToDo extends Test, Case {
  override predicate run(Qnit test) {
    if doubleQuoteEscaping("foo") = "\"foo\""
    then test.pass("Basic double quote escaping nothing to do works")
    else test.fail("Basic double quote escaping nothing to do doesn't work")
  }
}

class TestDoubleQuoteEscapingBasic extends Test, Case {
  override predicate run(Qnit test) {
    if doubleQuoteEscaping("foo\"bar") = "\"foo\\\"bar\""
    then test.pass("Basic double quote escaping works")
    else test.fail("Basic double quote escaping doesn't work")
  }
}

class TestDoubleQuoteEscapingWillEscapeTheEscaper extends Test, Case {
  override predicate run(Qnit test) {
    if doubleQuoteEscaping("foo\"bar\"baz") = "\"foo\\\"bar\\\"baz\""
    then test.pass("Double quote escaping will escape the escaper")
    else test.fail("Double quote escaping not properly escaping the escaper")
  }
}

class TestSingleQuoteEscapingNothingToDo extends Test, Case {
  override predicate run(Qnit test) {
    if singleQuoteEscaping("foo") = "'foo'"
    then test.pass("Basic single quote escaping nothing to do works")
    else test.fail("Basic single quote escaping nothing to do doesn't work")
  }
}

class TestUnescapeNothingToDo extends Test, Case {
  override predicate run(Qnit test) {
    if unescape("foo", "$") = "foo"
    then test.pass("Basic unescape nothing to do works")
    else test.fail("Basic unescape nothing to do doesn't work")
  }
}

class TestUnescapeBasic extends Test, Case {
  override predicate run(Qnit test) {
    if unescape("foo$bar", "$") = "foobar"
    then test.pass("Basic unescape works")
    else test.fail("Basic unescape doesn't work")
  }
}

class TestUnescapeWillUnescapeTheEscaper extends Test, Case {
  override predicate run(Qnit test) {
    if unescape("foo$$bar", "$") = "foo$bar"
    then test.pass("Unescape will unescape the escaper")
    else test.fail("Unescape not properly unescaping the escaper")
  }
}

class TestUnescapeWithManyEscapeCharacters extends Test, Case {
  override predicate run(Qnit test) {
    if unescape("$$$$$five $$$$four $$$three $$two $one", "$") = "$$five $$four $three $two one"
    then test.pass("Unescape with many escape characters works")
    else test.fail("Unescape with many escape characters doesn't work")
  }
}