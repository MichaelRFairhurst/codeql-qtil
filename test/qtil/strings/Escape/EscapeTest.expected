| All 17 tests passed. |
